pkg/apis/pubsub/v1alpha1/topic_types.go

```diff
@@ -0,0 +1,160 @@
+
+	// Spec defines the desired state of the Topic.
+	Spec TopicSpec `json:"spec,omitempty"`
+
+	// Status represents the current state of the Topic. This data may be out of
+	// date.
+	// +optional
+	Status TopicStatus `json:"status,omitempty"`
+}
+
+// Check that Topic can be validated, can be defaulted, and has immutable fields.
+var _ runtime.Object = (*Topic)(nil)
+var _ webhook.GenericCRD = (*Topic)(nil)
+
+// Check that Topic implements the Conditions duck type.
+var _ = duck.VerifyType(&Topic{}, &duckv1beta1.Conditions{})
+
+// TopicSpec defines an parameters for creating or publishing to a Cloud
+// Pub/Sub Topic depending the PropagationPolicy.
+type TopicSpec struct {
+	// Secret is the credential to be used to create and publish into the
+	// Cloud Pub/Sub Topic. The value of the secret entry must be a service
+	// account key in the JSON format
+	// (see https://cloud.google.com/iam/docs/creating-managing-service-account-keys).
+	Secret *corev1.SecretKeySelector `json:"secret,omitempty"`
+
+	// Project is the ID of the Google Cloud Project that the Pub/Sub
+	// Topic will be created in or used from.
+	Project string `json:"project,omitempty"`
+
+	// Topic is the ID of the Topic to create/use in Google Cloud Pub/Sub.
+	Topic string `json:"topic,omitempty"`
+
+	//PropagationPolicy defines how Topic controls the Cloud Pub/Sub topic for lifecycle changes.
+	PropagationPolicy PropagationPolicyType `json:"propagationPolicy,omitempty"`
+
+	// ServiceAccoutName is the name of the ServiceAccount that will be used to
+	// run the Topic Deployment.
+	ServiceAccountName string `json:"serviceAccountName,omitempty"`
+}
+
+// PropagationPolicyType defines enum type for TopicPolicy
+type PropagationPolicyType string
+
+const (
+	// TopicPolicyCreateDelete defines the Cloud Pub/Sub topic management
+	// policy for creating topic if not present, and delete topic when the
+	// Topic resource is deleted.
+	TopicPolicyCreateDelete PropagationPolicyType = "CreateDelete"
+
+	// TopicPolicyCreateRestrictDelete defines the Cloud Pub/Sub topic
+	// management policy for creating topic if not present, and do not delete
+	// topic when the Topic resource is deleted.
+	TopicPolicyCreateRestrictDelete PropagationPolicyType = "CreateRestrictDelete"
+
+	// TopicPolicyRestrictCreateRestrictDelete defines the Cloud Pub/Sub topic
+	// management policy for only using existing topics, and do not delete
+	// topic when the Topic resource is deleted.
+	TopicPolicyRestrictCreateRestrictDelete PropagationPolicyType = "RestrictCreateRestrictDelete"
+)
+
+var topicCondSet = apis.NewLivingConditionSet(
+	TopicConditionAddressable,
+	TopicConditionTopicExists,
+	TopicConditionPublisherReady,
+)
+
+const (
+	// TopicConditionReady has status True when all subconditions below have
+	// been set to True.
+	TopicConditionReady = apis.ConditionReady
+
+	// TopicConditionAddressable has status true when this Topic meets the
+	// Addressable contract and has a non-empty hostname.
```

Comment:
`hostname or URL`? Shouldn't we target the v1beta1 Addressable even if we support v1alpha1?